src/api/api.c

```diff
@@ -1106,3 +1162,116 @@ int8_t zp_send_join(z_session_t zs, const zp_send_join_options_t *options) {
+    _z_bytes_pair_list_t *current = this_->_inner;
+    while (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        if (_z_bytes_eq(&key, &head->key)) {
+            return _z_bytes_wrap(head->value.start, head->value.len);
+        }
+    }
+    return z_bytes_null();
+}
+void z_bytes_map_insert_by_alias(const z_owned_bytes_map_t *this_, z_bytes_t key, z_bytes_t value) {
+    _z_bytes_pair_list_t *current = this_->_inner;
+    while (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        if (_z_bytes_eq(&key, &head->key)) {
+            break;
+        }
+        current = _z_bytes_pair_list_tail(current);
+    }
+    if (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        _z_bytes_clear(&head->value);
+        head->value = _z_bytes_wrap(value.start, value.len);
+    } else {
+        struct _z_bytes_pair_t *insert = zp_malloc(sizeof(struct _z_bytes_pair_t));
+        memset(insert, 0, sizeof(struct _z_bytes_pair_t));
+        insert->key = _z_bytes_wrap(key.start, key.len);
+        insert->value = _z_bytes_wrap(value.start, value.len);
+        ((z_owned_bytes_map_t *)this_)->_inner = _z_bytes_pair_list_push(this_->_inner, insert);
+    }
+}
+void z_bytes_map_insert_by_copy(const z_owned_bytes_map_t *this_, z_bytes_t key, z_bytes_t value) {
+    _z_bytes_pair_list_t *current = this_->_inner;
+    while (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        if (_z_bytes_eq(&key, &head->key)) {
+            break;
+        }
+        current = _z_bytes_pair_list_tail(current);
+    }
+    if (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        _z_bytes_clear(&head->value);
+        _z_bytes_copy(&head->value, &value);
+        if (!head->key._is_alloc) {
+            _z_bytes_copy(&head->key, &key);
+        }
+    } else {
+        struct _z_bytes_pair_t *insert = zp_malloc(sizeof(struct _z_bytes_pair_t));
+        memset(insert, 0, sizeof(struct _z_bytes_pair_t));
+        _z_bytes_copy(&insert->key, &key);
+        _z_bytes_copy(&insert->value, &value);
+        ((z_owned_bytes_map_t *)this_)->_inner = _z_bytes_pair_list_push(this_->_inner, insert);
+    }
+}
+int8_t z_bytes_map_iter(const z_owned_bytes_map_t *this_, z_attachment_iter_body_t body, void *ctx) {
+    _z_bytes_pair_list_t *current = this_->_inner;
+    while (current) {
+        struct _z_bytes_pair_t *head = _z_bytes_pair_list_head(current);
+        int8_t ret = body(head->key, head->value, ctx);
+        if (ret) {
+            return ret;
+        }
+        current = _z_bytes_pair_list_tail(current);
+    }
+    return 0;
+}
+z_owned_bytes_map_t z_bytes_map_new(void) { return (z_owned_bytes_map_t){._inner = _z_bytes_pair_list_new()}; }
+z_owned_bytes_map_t z_bytes_map_null(void) { return (z_owned_bytes_map_t){._inner = NULL}; }
+z_bytes_t z_bytes_new(const char *str) {
```

Comment:
Why taking a `const char *str` as argument for a `z_bytes_t`? I would expect to take a `const uint8_t *bytes`.

Reply:
Mostly because this constructor is mostly aimed at constructing a borrowed slice from a null terminated string. We can remove it, or make it `uint8_t` instead, but since we never care about sign in anything `z_bytes_t` related, I think the `char*` constructor would be convenient for users.

Reply:
I agree with the rationale, then I'd call the function `z_bytes_from_str` or something like that. 
Regarding my comment, there is already an internal `_z_bytes_t _z_bytes_wrap(const uint8_t *bs, size_t len);` function that could be exposed for creating `z_bytes_t` from a `const uint8_t *bs`.

Reply:
I propose we expose `z_bytes_wrap` as a way to construct slices, and `z_bytes_from_str` for convenience (documenting that it is `strlen` based and therefore should only be run on trusted input).

`_z_bytes_make(size_t capacity)` _must_ remain private, as it's a way to construct an allocated slice, and breaks the ownership model we expose publicly.

Reply:
I agree with what you propose.